Sources/AsyncAlgorithms/AsyncChannel.swift

```diff
@@ -50,68 +53,49 @@ public final class AsyncChannel<Element: Sendable>: AsyncSequence, Sendable {
     }
   }
   
-  struct Awaiting: Hashable {
+  typealias Pending = ChannelToken<UnsafeContinuation<UnsafeContinuation<Element?, Never>?, Never>>
+  typealias Awaiting = ChannelToken<UnsafeContinuation<Element?, Never>>
+
+  struct ChannelToken<Continuation>: Hashable {
     var generation: Int
-    var continuation: UnsafeContinuation<Element?, Never>?
-    let cancelled: Bool
-    
-    init(generation: Int, continuation: UnsafeContinuation<Element?, Never>) {
+    var continuation: Continuation?
+
+    init(generation: Int, continuation: Continuation) {
       self.generation = generation
       self.continuation = continuation
-      cancelled = false
     }
-    
+
     init(placeholder generation: Int) {
       self.generation = generation
       self.continuation = nil
-      cancelled = false
     }
-    
-    init(cancelled generation: Int) {
-      self.generation = generation
-      self.continuation = nil
-      cancelled = true
-    }
-    
+
     func hash(into hasher: inout Hasher) {
       hasher.combine(generation)
     }
-    
-    static func == (_ lhs: Awaiting, _ rhs: Awaiting) -> Bool {
+
+    static func == (_ lhs: ChannelToken, _ rhs: ChannelToken) -> Bool {
       return lhs.generation == rhs.generation
     }
   }
+
+  enum ChannelTokenStatus: Equatable {
+    case new
+    case cancelled
+  }
   
   enum Emission {
     case idle
-    case pending([UnsafeContinuation<UnsafeContinuation<Element?, Never>?, Never>])
+    case pending(Set<Pending>)
     case awaiting(Set<Awaiting>)
```

Comment:
also the set of awaiting might pose issue too come to think of it... (I should have caught this earlier)